flink/src/main/java/org/apache/iceberg/flink/data/FlinkValueReaders.java

```diff
@@ -0,0 +1,314 @@
+    private static final StringReader INSTANCE = new StringReader();
+
+    private StringReader() {
+    }
+
+    @Override
+    public StringData read(Decoder decoder, Object reuse) throws IOException {
+      // use the decoder's readString(Utf8) method because it may be a resolving decoder
+      Utf8 utf8 = null;
+      if (reuse instanceof StringData) {
+        utf8 = new Utf8(((StringData) reuse).toBytes());
+      }
+
+      Utf8 string = decoder.readString(utf8);
+      return StringData.fromBytes(string.getBytes(), 0, string.getByteLength());
+    }
+  }
+
+  private static class EnumReader implements ValueReader<StringData> {
+    private final StringData[] symbols;
+
+    private EnumReader(List<String> symbols) {
+      this.symbols = new StringData[symbols.size()];
+      for (int i = 0; i < this.symbols.length; i += 1) {
+        this.symbols[i] = StringData.fromBytes(symbols.get(i).getBytes(StandardCharsets.UTF_8));
+      }
+    }
+
+    @Override
+    public StringData read(Decoder decoder, Object ignore) throws IOException {
+      int index = decoder.readEnum();
+      return symbols[index];
+    }
+  }
+
+  private static class DecimalReader implements ValueReader<DecimalData> {
+    private final ValueReader<byte[]> bytesReader;
+    private final int precision;
+    private final int scale;
+
+    private DecimalReader(ValueReader<byte[]> bytesReader, int precision, int scale) {
+      this.bytesReader = bytesReader;
+      this.precision = precision;
+      this.scale = scale;
+    }
+
+    @Override
+    public DecimalData read(Decoder decoder, Object reuse) throws IOException {
+      byte[] bytes = bytesReader.read(decoder, null);
+      return DecimalData.fromBigDecimal(new BigDecimal(new BigInteger(bytes), scale), precision, scale);
+    }
+  }
+
+  private static class TimeMicrosReader implements ValueReader<Integer> {
+    private static final TimeMicrosReader INSTANCE = new TimeMicrosReader();
+
+    @Override
+    public Integer read(Decoder decoder, Object reuse) throws IOException {
+      long micros = decoder.readLong();
+      // Flink only support time mills, just erase micros.
+      return (int) (micros / 1000);
+    }
+  }
+
+  private static class TimestampMillsReader implements ValueReader<TimestampData> {
+    private static final TimestampMillsReader INSTANCE = new TimestampMillsReader();
+
+    @Override
+    public TimestampData read(Decoder decoder, Object reuse) throws IOException {
+      // TODO Do we need to consider time zones.
+      return TimestampData.fromEpochMillis(decoder.readLong());
+    }
+  }
+
+  private static class TimestampMicrosReader implements ValueReader<TimestampData> {
+    private static final TimestampMicrosReader INSTANCE = new TimestampMicrosReader();
+
+    @Override
+    public TimestampData read(Decoder decoder, Object reuse) throws IOException {
+      // TODO Do we need to consider time zones.
```

Comment:
Time zones are left to the processing engine. It is up to the engine to convert times to concrete values for storage and from concrete values for display. Iceberg's responsibility is to return the value without modification.